### ISSUE TITLE
whereDate does not have consistent behaviour between MySQL and SQLite

### ISSUE THREAD

**Issue by vpratfr** [laravel/framework#24693]
- Laravel Version: latest 5.6
- PHP Version: 7.1
- Database Driver & Version: MySQL version bundled in Homestead, SQLite provided by PHP

### Description:

The following Eloquent query does not give the same result in unit tests and in homestead

```
$closedAt = Carbon::create(2017, 1, 31);
Model::whereDate('closed_at', '<', $closedAt)
                   ->orderByDesc('closed_at')
                   ->take(1)
                   ->first()
```

The following fixes the issue, but does not feel as right as it should be. whereDate is supposed to accept Carbon instances, right?

```
$closedAt = Carbon::create(2017, 1, 31);
Model::whereDate('closed_at', '<', $closedAt->format('Y-m-d'))
                   ->orderByDesc('closed_at')
                   ->take(1)
                   ->first()
```

I found the workaround when reading #10006. In that other issue, @jairovsky found the source of the issue (which is apparently caused to an issue in the SQLite Grammar). 

According to me, such a behaviour should either be :  

- forbidden (throw when passed a Carbon instance, for consistency, but not ideal at all)
- documented somewhere (maybe I missed it) 
- or better, fixed in the SQLite grammar.

**Comment by gdpa** [comment]
As i know you can use "where" instead of "whereDate" on eloquent models.

**Comment by vpratfr** [comment]
> As i know you can use "where" instead of "whereDate" on eloquent models.

Which does not do the same thing and does not resolve the aboce bug.

**Comment by kylekatarnls** [comment]
Hi, when you say "The following Eloquent query does not give the same result in unit tests and in homestead" can you give a concrete example: table sample, output in MySQL, output in SQLite?

**Comment by vpratfr** [comment]
Hi,

Sorry for the delay in answering, holiday time here in France...

Indeed, the DB field is created as a date: `$table->date('closed_at');`

I had a try to reproduce this inside a clean Laravel project but could not succeed. Hence I wonder if this has already been fixed or if that was some side effect of other code inside our project.

I'll close this for now and re-open if I can find a reliable way to reproduce.

**Comment by troygilbert** [comment]
I'm definitely encountering this issue, and like you, thought that passing Carbon instances into `whereDate()` would cast the Carbon instance to a date string, but instead it just calls Carbon's `__toString()` which results in a datetime string, and the SQL generated by `whereDate()` does not wrap the value in a `DATE()` function. But I think it should, right?

In fact, I thought I did, and I swear I relying on this assumption all over our codebase. But I found a bug today related to it (similar to what you described).
